src/harness/harness.c

```diff
@@ -36,49 +36,884 @@ tHarness_game_config harness_game_config;
+    0,     /* KEY_DELETE */
+    0,     /* KEY_HOME */
+    0,     /* KEY_END */
+    0,     /* KEY_PAGEUP */
+    0,     /* KEY_PAGEDOWN */
+    0,     /* KEY_LEFT */
+    0,     /* KEY_RIGHT */
+    0,     /* KEY_UP */
+    0,     /* KEY_DOWN */
+    0,     /* KEY_KP_NUMLOCK */
+    '/',   /* KEY_KP_DIVIDE */
+    '*',   /* KEY_KP_MULTIPLY */
+    '-',   /* KEY_KP_MINUS */
+    '+',   /* KEY_KP_PLUS */
+    '.',   /* KEY_KP_PERIOD */
+    0,     /* KEY_KP_EQUALS */
+    '0',   /* KEY_KP_0 */
+    '1',   /* KEY_KP_1 */
+    '2',   /* KEY_KP_2 */
+    '3',   /* KEY_KP_3 */
+    '4',   /* KEY_KP_4 */
+    '5',   /* KEY_KP_5 */
+    '6',   /* KEY_KP_6 */
+    '7',   /* KEY_KP_7 */
+    '8',   /* KEY_KP_8 */
+    '9',   /* KEY_KP_9 */
+    0,     /* KEY_F1 */
+    0,     /* KEY_F2 */
+    0,     /* KEY_F3 */
+    0,     /* KEY_F4 */
+    0,     /* KEY_F5 */
+    0,     /* KEY_F6 */
+    0,     /* KEY_F7 */
+    0,     /* KEY_F8 */
+    0,     /* KEY_F9 */
+    0,     /* KEY_F10 */
+    0,     /* KEY_F11 */
+    0,     /* KEY_F12 */
+    0,     /* KEY_PRTSCN */
+    0,     /* KEY_SCRLK */
+    0,     /* KEY_PAUSE */
+    ' ',   /* KEY_SPACE */
+};
+
 // German ASCII codes
 static int carmageddon_german_ascii_table[128] = {
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107,
-    108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 122, 121, 94, -33, -76, 8, 13, 13, 0, 45, 60, -10, -28, 46, 44, -4, 43, 35, 27,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -4, 56, -33, -76, 46, 0, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0,     /* KEY_SHIFT_ANY */
+    0,     /* KEY_ALT_ANY */
+    0,     /* KEY_CTRL_ANY */
+    0,     /* KEY_CTRL_ANY_2 */
+    0,     /* KEY_CAPSLOCK */
+    0,     /* KEY_RSHIFT */
+    0,     /* KEY_RALT */
+    0,     /* KEY_RCTRL */
+    0,     /* KEY_LSHIFT */
+    0,     /* KEY_LALT */
+    0,     /* KEY_LCTRL */
+    '0',   /* KEY_0 */
+    '1',   /* KEY_1 */
+    '2',   /* KEY_2 */
+    '3',   /* KEY_3 */
+    '4',   /* KEY_4 */
+    '5',   /* KEY_5 */
+    '6',   /* KEY_6 */
+    '7',   /* KEY_7 */
+    '8',   /* KEY_8 */
+    '9',   /* KEY_9 */
+    'a',   /* KEY_A */
+    'b',   /* KEY_B */
+    'c',   /* KEY_C */
+    'd',   /* KEY_D */
+    'e',   /* KEY_E */
+    'f',   /* KEY_F */
+    'g',   /* KEY_G */
+    'h',   /* KEY_H */
+    'i',   /* KEY_I */
+    'j',   /* KEY_J */
+    'k',   /* KEY_K */
+    'l',   /* KEY_L */
+    'm',   /* KEY_M */
+    'n',   /* KEY_N */
+    'o',   /* KEY_O */
+    'p',   /* KEY_P */
+    'q',   /* KEY_Q */
+    'r',   /* KEY_R */
+    's',   /* KEY_S */
+    't',   /* KEY_T */
+    'u',   /* KEY_U */
+    'v',   /* KEY_V */
+    'w',   /* KEY_W */
+    'x',   /* KEY_X */
+    'z',   /* KEY_Y */
+    'y',   /* KEY_Z */
+    '^',   /* KEY_GRAVE */
+    -0x21, /* KEY_MINUS */
```

Comment:
Is it supposed to be `-0x21`?

Reply:
I think this is a German Eszett (ẞ).
```python
b'\xdf'.decode("windows-1252")
```
The binary really contains `FFFFFFDFh`.
I think their compiler treated characters as signed ==> the 8-bit `0xdf` value is negative and must be sign-extended to a 32-bit integer.

Reply:
The German DOS version contains the same codes, but they are all positive there.
So it looks like this is an oversight/bug of OG.

The text input behavior of German Dethrace and the German DOS version is now the same.
I didn't verify the German Windows executable.